### PR TITLE
GC: fix potential compactor issue that it may not have all serializers.

### PR DESCRIPTION
**Pull request by WenbinZhu**
# Overview

Log unit server might not have all serializers of different streams, but
the current compaction process deserializes all entries which could fail.
However compactor actually does not need to get the exact deserialized
content, so this patch adds a support to only deserialize the entry into
a list of byte arrays and re-assemble after compaction.

Related issue(s) (if applicable): #2235 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java [hunk 1]
```diff
@@ -28,6 +28,8 @@
 
     private ByteBuf serializedCache = null;
 
+    // last know serialized payload size.
+    @Getter
     private int lastKnownSize = NOT_KNOWN;
```
Comment (lixinchengdu) [inline runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java]: Add comments about this variable and also states your concern mentioned in the lasted issue.

File: runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRRecord.java [hunk 2]
```diff
@@ -193,8 +212,33 @@ void serialize(ByteBuf b) {
...
+    }
+
+    /**
+     * Get the bytes of one {@link SMRRecord} from a larger byte buffer
+     * and transfer it to a new byte array.
+     *
+     * @param b byte buffer to transfer from
+     * @return a new byte array containing the transferred bytes
+     */
+    static byte[] transfer(ByteBuf b) {
```
Comment (lixinchengdu) [inline runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRRecord.java]: This name is confusing. Probably toByte is a better name.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java (hunk 1 above)
Comment (WenbinZhu) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java]: There is a comment on this variable.
As for the concern in that issue, I addressed it in this PR. I will reference it in PR description.

File: runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRRecord.java (hunk 2 above)
Comment (WenbinZhu) [reply runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRRecord.java]: Changed name to slice, as it's not only converting to bytes, it slices from a larger buffer.

**Pull request merged**